Signal/src/call/CallAudioService.swift

```diff
@@ -0,0 +1,177 @@
+
+    // MARK: Vibration config
+    private let vibrateRepeatDuration = 1.6
+
+    // Our ring buzz is a pair of vibrations.
+    // `pulseDuration` is the small pause between the two vibrations in the pair.
+    private let pulseDuration = 0.2
+
+    public var isSpeakerphoneEnabled = false {
+        didSet {
+            handleUpdatedSpeakerphone()
+        }
+    }
+
+    // MARK: - Initializers
+
+    init(handleRinging: Bool) {
+        self.handleRinging = handleRinging
+    }
+
+    // MARK: - Service action handlers
+
+    public func handleState(_ state: CallState) {
+        switch state {
+        case .idle: handleIdle()
+        case .dialing: handleDialing()
+        case .answering: handleAnswering()
+        case .remoteRinging: handleRemoteRinging()
+        case .localRinging: handleLocalRinging()
+        case .connected: handleConnected()
+        case .localFailure: handleLocalFailure()
+        case .localHangup: handleLocalHangup()
+        case .remoteHangup: handleRemoteHangup()
+        case .remoteBusy: handleBusy()
+        }
+    }
+
+    private func handleIdle() {
+        Logger.debug("\(TAG) \(#function)")
+    }
+
+    private func handleDialing() {
+        Logger.debug("\(TAG) \(#function)")
+    }
+
+    private func handleAnswering() {
+        Logger.debug("\(TAG) \(#function)")
+        stopRinging()
+    }
+
+    private func handleRemoteRinging() {
+        Logger.debug("\(TAG) \(#function)")
+    }
+
+    private func handleLocalRinging() {
+        Logger.debug("\(TAG) in \(#function)")
+        startRinging()
+    }
+
+    private func handleConnected() {
+        Logger.debug("\(TAG) \(#function)")
+        stopRinging()
+
+        // disable start recording to transmit call audio.
+        setAudioSession(category: AVAudioSessionCategoryPlayAndRecord)
+    }
+
+    private func handleLocalFailure() {
+        Logger.debug("\(TAG) \(#function)")
+        stopRinging()
+    }
+
+    private func handleLocalHangup() {
+        Logger.debug("\(TAG) \(#function)")
+        stopRinging()
+    }
+
+    private func handleRemoteHangup() {
+        Logger.debug("\(TAG) \(#function)")
+        stopRinging()
+    }
+
+    private func handleBusy() {
+        Logger.debug("\(TAG) \(#function)")
+        stopRinging()
+    }
+
+    private func handleUpdatedSpeakerphone() {
+        if isSpeakerphoneEnabled {
+            setAudioSession(category: AVAudioSessionCategoryPlayAndRecord, options: .defaultToSpeaker)
+        } else {
+            setAudioSession(category: AVAudioSessionCategoryPlayAndRecord)
+        }
+    }
+
+    // MARK: - Ringing
+
+    private func startRinging() {
+        guard handleRinging else {
+            Logger.debug("\(TAG) ignoring \(#function) since CallKit handles it's own ringing state")
+            return
+        }
+
+        vibrateTimer = Timer.scheduledTimer(timeInterval: vibrateRepeatDuration, target: self, selector: #selector(ringVibration), userInfo: nil, repeats: true)
```

Comment:
`NSTimer`/`Timer` is a classic source of leaks, because it holds a strong reference to `target`.  I think that'll bite us here.  In the past, what I've done is add a NSTimer category with methods that take a weak reference.